File: src/ssh_manager/util/secrets.py
"""Secret resolution - keep provider tokens out of plaintext ``.env`` if you want.

A token env var (e.g. ``GH_TOKEN``) normally holds the secret directly. To avoid
storing it at rest, set the value to ``cmd:<command>`` and the command is run at
use-time; its stdout (trimmed) is the secret. The command is split with shell
word rules (``shlex``) and run via the argv-only subprocess chokepoint (never
``shell=True``) - so it integrates with any secret manager:

    GH_TOKEN=cmd:op read op://Private/GitHub/token        # 1Password CLI
    GH_TOKEN=cmd:keyring get ssh-manager gh_token              # OS keyring
    GH_TOKEN=cmd:age -d -i ~/.config/ssh-manager/age-identity.txt ~/gh.token.age

A plain value is returned as-is, so this is fully backward compatible.
"""
from __future__ import annotations

import shlex
from functools import lru_cache

from . import proc

_CMD_PREFIX = "cmd:"


@lru_cache(maxsize=128)
def _run_cmd_secret(command: str) -> str | None:
    """Run a ``cmd:`` secret command once and cache its result for the process.

    Memoized because a single provider operation resolves the token several times
    (capability check, env, then the CLI call) - without caching, a slow or
    biometric-prompting secret manager would fire 2-3x per host. A failure / timeout
    / empty output yields ``None`` (the provider degrades to the manual path)."""
    argv = shlex.split(command)
    if not argv:
        return None
    result = proc.run(argv, timeout=20)
    if result.returncode != 0:
        return None
    return result.stdout.strip() or None


def resolve_secret(raw: str | None) -> str | None:
    """Resolve a token value: ``cmd:<command>`` runs the command and returns its
    trimmed stdout (memoized per process); anything else is returned unchanged.
    ``None``/empty -> ``None``."""
    if not raw:
        return None
    if not raw.startswith(_CMD_PREFIX):
        return raw
    return _run_cmd_secret(raw[len(_CMD_PREFIX):])

File: internal/util/secrets/secrets.go
// Package secrets resolves provider tokens, ported from util/secrets.py. A token
// value of cmd:<command> runs the command at use-time and uses its trimmed stdout
// as the secret (so it integrates with any secret manager); anything else is
// returned as-is. Results are memoized per process.
package secrets

import (
	"os/exec"
	"strings"
	"sync"
)

const cmdPrefix = "cmd:"

var cache sync.Map // command -> resolved secret ("" == failure/empty)

// Resolve returns the secret for a token value: cmd:<command> runs the command
// (memoized), anything else is returned unchanged; "" stays "".
func Resolve(raw string) string {
	if raw == "" {
		return ""
	}
	if !strings.HasPrefix(raw, cmdPrefix) {
		return raw
	}
	command := raw[len(cmdPrefix):]
	if v, ok := cache.Load(command); ok {
		return v.(string)
	}
	out := runCmdSecret(command)
	cache.Store(command, out)
	return out
}

func runCmdSecret(command string) string {
	argv := shlexSplit(command)
	if len(argv) == 0 {
		return ""
	}
	out, err := exec.Command(argv[0], argv[1:]...).Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// shlexSplit is a minimal POSIX shell-word splitter (handles single/double quotes
// and backslash escapes) - enough for cmd: secret commands. Mirrors shlex.split.
func shlexSplit(s string) []string {
	var args []string
	var cur strings.Builder
	inWord := false
	quote := byte(0) // 0, '\'', or '"'
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case quote != 0:
			if c == quote {
				quote = 0
			} else if quote == '"' && c == '\\' && i+1 < len(s) {
				i++
				cur.WriteByte(s[i])
			} else {
				cur.WriteByte(c)
			}
		case c == '\'' || c == '"':
			quote = c
			inWord = true
		case c == '\\' && i+1 < len(s):
			i++
			cur.WriteByte(s[i])
			inWord = true
		case c == ' ' || c == '\t' || c == '\n':
			if inWord {
				args = append(args, cur.String())
				cur.Reset()
				inWord = false
			}
		default:
			cur.WriteByte(c)
			inWord = true
		}
	}
	if inWord {
		args = append(args, cur.String())
	}
	return args
}
